implementations/rust/ockam/ockam/src/system/builder.rs

```diff
@@ -0,0 +1,296 @@
+        let prev_id = prev_id.into();
+        if let Some(ref mut prev) = self.inner.get_mut(&prev_id) {
+            prev.routes.insert(rule.into(), addr.clone());
+        }
+        self.add(addr, id, handler)
+    }
+
+    /// Re-address every system handler in order to clone it
+    ///
+    /// When initialising a worker system builder you setup handlers
+    /// and routes between them.  However when cloning the worker
+    /// system all addresses need to be re-addressed in order to allow
+    /// a second instance of every address to exist on the same node.
+    ///
+    /// This problem may be solved by having scoped addresses in the
+    /// future.
+    ///
+    /// This function assigns a new address to every handler and then
+    /// changes routes that point to the old addresses.  Any route
+    /// that is not in the internal set (i.e. the fin address routes)
+    /// are replaced with the externally provided fin-address.
+    pub fn readdress(&mut self, fin_addr: &Address) {
+        // Generate a lookup-table first with new addresses to replace
+        // the old addresses
+        let addrs: BTreeMap<_, _> = self
+            .inner
+            .iter()
+            .map(|(_, data)| (data.addr.clone(), Address::random(0)))
+            .collect();
+
+        // Small utility function that will either return the the new
+        // address if it is within the system, or the new provided fin
+        // address if it external (this only supports _one_ fin
+        // endpoint)
+        fn get_addr(
+            addrs: &BTreeMap<Address, Address>,
+            prev: &Address,
+            fin_addr: &Address,
+        ) -> Address {
+            addrs.get(prev).unwrap_or(fin_addr).clone()
+        }
+
+        // Fix the entry point
+        if let Some(entry) = core::mem::replace(&mut self.entry, None) {
+            self.entry = addrs.get(&entry).cloned();
+        }
+
+        self.inner.iter_mut().for_each(|(_, data)| {
+            data.addr = get_addr(&addrs, &data.addr, fin_addr);
+            data.routes.iter_mut().for_each(|(_id, route_addr)| {
+                *route_addr = get_addr(&addrs, route_addr, fin_addr);
+            });
+        });
+    }
+
+    /// Create a `WorkerSystem` and pre-initialise every SystemHandler
+    pub async fn finalise(self, ctx: &mut C) -> Result<WorkerSystem<C, M>> {
+        let mut system = WorkerSystem::default();
+        if let Some(addr) = self.entry {
+            system.set_entry(addr);
+        }
+
+        for (
+            _identifier,
+            HandlerData {
+                addr,
+                mut inner,
+                mut routes,
+            },
+        ) in self.inner
+        {
+            inner.initialize(ctx, &mut routes).await?;
+            system.attach_boxed(addr, inner);
+        }
+
+        Ok(system)
+    }
+}
+
+/// Builder API for a single SystemHandler
+///
+/// This type may panic during destruction if not properly initialised
```

Comment:
Not a fan of this in general, but the work to do in a typestatey way is not worth it, and having a `fn finish(self) -> Result<()>` is also not good (because its an API usage error).

So I suppose this is fine in this case.